### PR TITLE
schemachanger: handle DROP following CREATE in-txn

### LINKED ISSUE #104123: schemachanger: Add elements that were previously dropped
The motivation of this issue is that, today, declarative schema changer does not handle the following case well
```
SET use_declarative_schema_changer = 'unsafe_always';
BEGIN;
CREATE SCHEMA sc;
DROP SCHEMA sc;
.... -- more DDL stmts
COMMIT
```

Namely, `DROP SCHEMA` will need to revert target status of elements previously added from `CREATE SCHEMA`. What makes the issue more interesting is that those elements might have transitioned from their initial status during the statementPhase of `CREATE SCHEMA` and thus had side-effects on the (in-memory) desc collection. Ideally, those sides effects need to be un-done during the statementPhase of the `CREATE SCHEMA`.

### PR DESCRIPTION
**Pull request by postamar**
This commit fixes a few bugs which prevented the following kind of schema change from performing correctly:

    BEGIN;
    CREATE SCHEMA sc;
    DROP SCHEMA sc;

Informs #104123.

Release note: None

----

  scplan: remove hard-coded 1-op-edge constraint on statement phase
  
  Previously, the declarative schema changer had this arbitrary constraint
  which enforced that each element could transition at most once
  towards its target during the statement phase(s) in the statement
  transaction.
  
  This commit removes this and relies on the two-version-invariant rules
  instead. To do so, this commit reintroduces the
  PreviousTransactionPrecedence dependency edge kind and expresses those
  rules using that instead of the PreviousStagePrecedence kind, which is
  now deprecated.
  
  Removing this limitation is necessary for the side-effects of CREATE
  statements to be visible inside the statement transaction. This change
  does not otherwise significantly alter the plans which are generated.
  
  Informs #104351.
  Informs #104123.
  
  Release note: None

----

  scplan: stop generating pre-commit main stage if redundant
  
  This commit does not change any functionality in any significant way, it
  only makes the EXPLAIN (DDL) output nicer in cases like:
  
      BEGIN;
      CREATE SCHEMA sc;
      EXPLAIN (DDL) DROP SCHEMA sc;
  
  This enables us to test that the following performs correctly:
  
      BEGIN;
      CREATE SCHEMA sc;
      DROP SCHEMA sc;
      CREATE SCHEMA sc;
      COMMIT;
  
  Fixes #104123.
  
  Release note: None

### CODE REVIEW

File: pkg/sql/schemachanger/scplan/testdata/create_schema [hunk 1]
```diff
@@ -1,9 +1,9 @@
 ops
 CREATE SCHEMA sc;
 ----
-StatementPhase stage 1 of 1 with 7 MutationType ops
+StatementPhase stage 1 of 1 with 8 MutationType ops
   transitions:
-    [[Schema:{DescID: 104}, PUBLIC], ABSENT] -> DESCRIPTOR_ADDED
+    [[Schema:{DescID: 104}, PUBLIC], ABSENT] -> PUBLIC
```
Comment (fqazi) [inline pkg/sql/schemachanger/scplan/testdata/create_schema:6]: Nice!

**Review by fqazi — APPROVED** [review]
Thanks for doing this @postamar! Can we add test coverage for CREATE SCHMEA + DROP SCHEMA in a single transaction for planning or do we need something else

**Comment by postamar** [comment]
As it turns out, 
```
  BEGIN;
  CREATE SCHEMA sc;
  DROP SCHEMA sc;
  CREATE SCHEMA sc;
  COMMIT;
```
now works correctly.

File: pkg/sql/schemachanger/scplan/internal/scstage/build.go [hunk 2]
```diff
@@ -503,7 +499,32 @@ func (sb stageBuilder) isUnmetInboundDep(de *scgraph.DepEdge) bool {
...
+		}
+		// At this point, the current stage is in the statement transaction.
+		// The only way this inbound dependency can be met is if the source node
+		// status is the initial status before the schema change even began,
+		// in which case the node was fulfilled before the statement transaction.
+		idx, ok := sb.bc.targetToIdx[de.From().Target]
+		if !ok {
+			panic(errors.AssertionFailedf("unknown target %q", de.From().Target))
+		}
+		return de.From().CurrentStatus != sb.bc.initial[idx]
```
Comment (Xiang-Gu) [inline pkg/sql/schemachanger/scplan/internal/scstage/build.go:522]: @postamar I hope this is not too cold. I have some questions: This function essentially answers the question "is node `n` currently un-schedulable at this stage due to this inbound PreviousTransactionPrecedence edge?"

Your first half of the answer makes sense to me but the second, highlighted half does not. In particular, if `e.from()` is already fulfilled and we are in StatementPhase or PrecommitPhase, why do we need to consult whether `e.from()` is in its initial status to determine whether we return true of false here? 

In other words, if `de.from()` is currently *not* at its initial status, this means it must have transitioned in the StatementPhase of some previous statement, but how is that count as "previous transaction"? 
I want to know the precious definition of "previous transaction" in "PreviousTransactionPrecedence".

Comment (postamar) [reply pkg/sql/schemachanger/scplan/internal/scstage/build.go:522]: I think you've answered your own question in your 3rd paragraph: it's precisely because we're still in the same transaction that the dependency cannot be met (we're not allowed to visit `de.To()` in the same transaction as `de.From()`).
